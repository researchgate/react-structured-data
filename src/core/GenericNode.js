import JSONLDAbstractNode from "../JSONLDAbstractNode";

class GenericNode extends JSONLDAbstractNode {
  getJSON({ type, jsonldtype, ...schema }) {
    const parseChildren = super.parseChildren();
    const details = {
      "@type": jsonldtype || this.getSchemaOrgType(),
      ...schema
    };
    return type
      ? Object.assign({ [type]: details }, ...parseChildren)
      : Object.assign(details, ...parseChildren);
  }
}

export default GenericNode;
