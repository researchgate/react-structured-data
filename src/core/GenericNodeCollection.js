import JSONLDAbstractNode from "../JSONLDAbstractNode";
import T from "prop-types";

class GenericNodeCollection extends JSONLDAbstractNode {
  getJSON() {
    if (this.props.children) {
      return super.parseChildren();
    }
    return null;
  }
}

GenericNodeCollection.propTypes = {
  type: T.string.isRequired
};

export default GenericNodeCollection;
