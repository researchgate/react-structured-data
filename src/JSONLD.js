import React from "react";
import PropTypes from "prop-types";
import GenericNodeCollection from "./core/GenericNodeCollection";

export const JSONLD = props => {
  let json = null;
  if (props.children) {
    const ChildClass = props.children.type;
    const { children, type, ...schema } = props.children.props;

    const childClassInstance = new ChildClass(props.children.props);
    const firstChild = childClassInstance.getJSON(schema);

    json =
      childClassInstance instanceof GenericNodeCollection
        ? Object.assign(
            { "@context": "https://schema.org/" },
            { [type]: firstChild }
          )
        : Object.assign({ "@context": "https://schema.org/" }, firstChild);
  }

  return props.dangerouslyExposeHtml ? (
    <script
      type="application/ld+json"
      dangerouslySetInnerHTML={{ __html: JSON.stringify(json) }}
    />
  ) : (
    <script type="application/ld+json">{JSON.stringify(json)}</script>
  );
};

JSONLD.propTypes = {
  additionalType: PropTypes.object,
  dangerouslyExposeHtml: PropTypes.bool,
  children: PropTypes.node
};

JSONLD.defaultProps = {
  dangerouslyExposeHtml: false
};

export default JSONLD;
