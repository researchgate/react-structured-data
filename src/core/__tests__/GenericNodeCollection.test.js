import React from "react";
import renderer from "react-test-renderer";
import JSONLD from "../../JSONLD";
import GenericNode from "../GenericNode";
import GenericNodeCollection from "../GenericNodeCollection";

describe("GenerticNodeCollection", () => {
  it("filters out null and false nodes from a collection", () => {
    expect(
      renderer.create(
        <JSONLD>
          <GenericNodeCollection>
            <GenericNode type="author" jsonldtype="Person" name="Cool Carl" />
            {false}
            {null}
            <GenericNode jsonldtype="AdministrativeArea" name="Chicago, IL" />
          </GenericNodeCollection>
        </JSONLD>
      )
    ).toMatchSnapshot();
  });
});
