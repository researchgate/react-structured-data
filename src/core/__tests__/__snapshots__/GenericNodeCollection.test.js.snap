// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenerticNodeCollection filters out null and false nodes from a collection 1`] = `
<script
  type="application/ld+json"
>
  {"@context":"https://schema.org/","undefined":[{"author":{"@type":"Person","name":"Cool Carl"}},{"@type":"AdministrativeArea","name":"Chicago, IL"}]}
</script>
`;
